src/engine/thread.js

```diff
@@ -80,7 +99,14 @@ Thread.prototype.pushStack = function (blockId) {
     // Push an empty stack frame, if we need one.
     // Might not, if we just popped the stack.
     if (this.stack.length > this.stackFrames.length) {
+        // Copy warp mode from any higher level.
+        var warpMode = false;
+        if (this.stackFrames[this.stackFrames.length - 1]) {
+            warpMode = this.stackFrames[this.stackFrames.length - 1].warpMode;
+        }
         this.stackFrames.push({
+            isLoop: false, // Whether this level of the stack is a loop.
+            warpMode: warpMode, // Whether this level is in warp mode.
```

Comment:
Yeah, it would be nice to reduce this complexity. A function argument might work, but I'm a little nervous about making sure we always call with the right warpMode value...it would be worth trying to implement it that way but I'm not confident it'll be simpler. Also, when the thread runs out of warp time, we need to make sure when we resume execution of it, warp mode continues. Which might be annoying if we're not storing the value on the stack - we'll need to walk up to the proc definition again and look up the flag.

Hmmm...